### ISSUE TITLE
I2I: Remove AMP CORS protocol

### ISSUE THREAD

**Issue by dvoytenko** [ampproject/amphtml#22136]
## Summary

[AMP CORS](https://github.com/ampproject/amphtml/blob/master/spec/amp-cors-requests.md) was necessary in a pre-separated domain context. The transition to the fully separated origins has been done long time ago and deployments of AMP with a unified are no longer considered valid and easily detected on the server side.

## Design document

Thus the proposal is to remove most of response requirements for AMP CORS. Specifically, `AMP-Access-Control-Allow-Source-Origin` and `Access-Control-Expose-Headers` response headers will no longer be required. We will continue sending `__amp_source_origin` for backward compatibility for a foreseeable future.

## Motivation

This is an extra code where it's no longer needed. Additionally, the blanket cross-origin=* requests will be open to be used in AMP by default.

/cc @ampproject/wg-approvers
/cc @jridgewell @cramforce @Gregable @lannka

**Comment by lannka** [comment]
Happy to revisit but we were stalled on this last time (#14789) because the domain separation does not separate HTTPS & HTTP.

**Comment by cramforce** [comment]
@dvoytenko Where did we land on only removing the requirement for HTTPS origin documents?

**Comment by dvoytenko** [comment]
@lannka I wonder how this point changes given that we allow CORS to only https resources.

**Comment by dvoytenko** [comment]
@aghassemi after more research for `Access-Control-Allow-Origin`:

* Wildcards `*` within an origin are not allowed. It's either `*` or a valid origin.
* Wildcards indeed are not allowed for credentialed CORS.

Thus, I think, we can do a wholesome AMP CORS deprecation.

**Comment by choumx** [comment]
@sebastianbenz Once this lands, we should update [amp-toolbox/amp-cors](https://github.com/ampproject/amp-toolbox/tree/master/packages/cors) too.

**Comment by dvoytenko** [comment]
@choumx @sebastianbenz this has landed.

**Comment by sebastianbenz** [comment]
Whoop whoop! I'll deprecate amp-toolbox/amp-cors.

**Comment by ithinkihaveacat** [comment]
@sebastianbenz You mean update? The module injects still-necessary (but not AMP specific) headers such as `Access-Control-Allow-Origin` so sites that currently use it still need to generate these headers in some way. (Though they could eventually switch to e.g. https://expressjs.com/en/resources/middleware/cors.html.)

**Comment by sebastianbenz** [comment]
@ithinkihaveacat you're right, deprecation is not needed. Injecting `Access-Control-Allow-Origin` if the `__source_origin` header is present still makes sense.

**Comment by mattwomple** [comment]
When this I2I was brought up for design review, was there any discussion about the `AMP-Same-Origin` header? I haven't been able to glean any security benefit to that header either.

**Comment by dvoytenko** [comment]
No. It was not subject to this particular change. I believe `AMP-Same-Origin` header is used as a "proof" of same-site access since, unfortunately, forms do not implement CORS protocol on POST. There's probably some reason why that's still the case for forms, but it has always eluded me.

**Comment by mattwomple** [comment]
Ah, thanks for the response and clarification @dvoytenko!

**Comment by choumx** [comment]
Stumbled upon this again. FWIW, it looks like Firefox 70 finally extended `Origin` header to forms for CSRF mitigation: https://bugs.chromium.org/p/chromium/issues/detail?id=641620#c4

I believe this CSRF mitigation was the "origin" 🤓 of `AMP-Same-Origin`, so maybe we can get rid of it soon.
